src/executioners/OptimizeSolve.C

```diff
@@ -263,16 +259,58 @@ OptimizeSolve::objectiveAndGradientFunctionWrapper(
 }
 
 PetscErrorCode
-OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat hessian, Mat /*pc*/, void * ctx)
+OptimizeSolve::hessianFunctionWrapper(Tao /*tao*/, Vec x, Mat /*hessian*/, Mat /*pc*/, void * ctx)
 {
-  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  // this will need to set the real parameters (Vec x) on the solver (see ln 250) so that the
+  // current parameter can be used in the applyHessianWrapper.  This is not important for force
+  // inversion because the hessian does not change.  This will be important for material
+  // inversion.
 
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
   libMesh::PetscVector<Number> param(x, solver->_my_comm);
   *solver->_parameters = param;
+  auto n = solver->_ndof;
+  auto comm = solver->_my_comm.get();
+  MatCreateShell(
+      comm, n, n, n, n, ctx, &(solver->_hessian)); // need to fix two of the n's for parallel
+  MatShellSetOperation(
+      solver->_hessian, MATOP_MULT, (void (*)(void))OptimizeSolve::applyHessianWrapper);
+  return 0;
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessianWrapper(Mat H, Vec s, Vec Hs)
+{
+  void * ctx;
+  MatShellGetContext(H, &ctx);
+  auto * solver = static_cast<OptimizeSolve *>(ctx);
+  libMesh::PetscVector<Number> sbar(s, solver->_my_comm);
+  libMesh::PetscVector<Number> Hsbar(Hs, solver->_my_comm);
+  return solver->applyHessian(sbar, Hsbar);
+}
+
+PetscErrorCode
+OptimizeSolve::applyHessian(libMesh::PetscVector<Number> & s, libMesh::PetscVector<Number> & Hs)
```

Comment:
Could you move this function below `gradientFunction`? Feel like it makes more sense to put there.

Reply:
Done.